logstash-core/src/main/java/org/logstash/plugins/codecs/Line.java

```diff
@@ -0,0 +1,128 @@
+package org.logstash.plugins.codecs;
+
+import org.logstash.Event;
+import org.logstash.StringInterpolation;
+import co.elastic.logstash.api.Codec;
+import co.elastic.logstash.api.Configuration;
+import co.elastic.logstash.api.LogstashPlugin;
+import co.elastic.logstash.api.Context;
+import co.elastic.logstash.api.PluginHelper;
+import co.elastic.logstash.api.PluginConfigSpec;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CoderResult;
+import java.nio.charset.CodingErrorAction;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+
+@LogstashPlugin(name = "line")
+public class Line implements Codec {
+
+    public static final String DEFAULT_DELIMITER = "\n";
+
+    private static final PluginConfigSpec<String> CHARSET_CONFIG =
+            Configuration.stringSetting("charset", "UTF-8");
+
+    private static final PluginConfigSpec<String> DELIMITER_CONFIG =
+            Configuration.stringSetting("delimiter", DEFAULT_DELIMITER);
+
+    private static final PluginConfigSpec<String> FORMAT_CONFIG =
+            Configuration.stringSetting("format");
+
+    static final String MESSAGE_FIELD = "message";
+
+    private final String delimiter;
+    private final Charset charset;
+    private String format = null;
+
+    private final CharBuffer charBuffer = ByteBuffer.allocateDirect(64 * 1024).asCharBuffer();
+    private final CharsetDecoder decoder;
+    private String remainder = "";
+
+    public Line(final Configuration configuration, final Context context) {
+        delimiter = configuration.get(DELIMITER_CONFIG);
+        charset = Charset.forName(configuration.get(CHARSET_CONFIG));
+        format = configuration.get(FORMAT_CONFIG);
+        decoder = charset.newDecoder();
+        decoder.onMalformedInput(CodingErrorAction.IGNORE);
+    }
+
+    @Override
+    public void decode(ByteBuffer buffer, Consumer<Map<String, Object>> eventConsumer) {
+        int bufferPosition = buffer.position();
+        CoderResult result = decoder.decode(buffer, charBuffer, false);
+        charBuffer.flip();
+        String s = (remainder == null ? "" : remainder) + charBuffer.toString();
+        charBuffer.clear();
+
+        if (s.endsWith(delimiter)) {
+            // strip trailing delimiter, if any, to match Ruby implementation
+            s = s.substring(0, s.length() - delimiter.length());
+        } else {
+            int lastIndex = s.lastIndexOf(delimiter);
```

Comment:
possible bug? I'm not 100% sure I'm reading correctly, but I think that if the delimiter occurs multiple times in the buffer, we split on the _last_ delimiter, causing those multiples to be emitted together as a single event.

Reply:
nope; I totally read it wrong; we split on the delimiter later on. d'oh.